packages/apollo-server/src/utils/ApolloServer.ts

```diff
@@ -0,0 +1,383 @@
+import { ApolloEngine as Engine } from 'apollo-engine';
+import {
+  SubscriptionServer,
+  ExecutionParams,
+} from 'subscriptions-transport-ws';
+
+import { CorsOptions } from 'cors';
+import {
+  Config,
+  ListenOptions,
+  MiddlewareOptions,
+  MiddlewareRegistrationOptions,
+  ServerInfo,
+  Context,
+  ContextFunction,
+} from './types';
+
+export {
+  LogFunction,
+  ValidationContext,
+  CacheControlExtensionOptions,
+  GraphQLResolveInfo,
+};
+
+// taken from engine
+export function joinHostPort(host: string, port: number) {
+  if (host.includes(':')) host = `[${host}]`;
+  return `${host}:${port}`;
+}
+
+const env = process.env.NODE_ENV;
+const isDev = env !== 'production' && env !== 'test';
+
+const NoIntrospection = (context: ValidationContext) => ({
+  Field(node: FieldDefinitionNode) {
+    if (node.name.value === '__schema' || node.name.value === '__type') {
+      context.reportError(
+        new GraphQLError(
+          'GraphQL introspection is not allowed by Apollo Server, but the query containted __schema or __type. To enable introspection, pass introspection: true to ApolloServer in production',
+          [node],
+        ),
+      );
+    }
+  },
+});
+
+export class ApolloServerBase<
+  Server = HttpServer,
+  Request = any,
+  Cors = CorsOptions
+> {
+  app?: Server;
+  schema: GraphQLSchema;
+  private context?: Context | ContextFunction;
+  private engine?: Engine;
+  private appCreated: boolean = false;
+  private middlewareRegistered: boolean = false;
+  private http?: HttpServer;
+  private subscriptions?: any;
+  private graphqlEndpoint: string = '/graphql';
+  private cors?: Cors;
+  private engineEnabled: boolean = false;
+  private requestOptions: Partial<GraphQLOptions<any>>;
+  private disableTools: boolean = !isDev;
+
+  constructor(config: Config<Server, Request, Cors>) {
+    const {
+      app,
+      context,
+      cors,
+      engine,
+      engineInRequestPath,
+      resolvers,
+      schema,
+      schemaDirectives,
+      subscriptions,
+      typeDefs,
+      introspection,
+      ...requestOptions
+    } = config;
+
+    if (introspection === false || !isDev) {
+      this.disableTools = true;
+      const noIntro = [NoIntrospection];
+      requestOptions.validationRules = requestOptions.validationRules
+        ? requestOptions.validationRules.concat(noIntro)
+        : noIntro;
+    }
+
+    this.requestOptions = requestOptions;
+    this.context = context;
+    // XXX should we move this to the `start` call? This would make hot
+    // reloading eaiser but may not be worth it?
+    this.schema = schema
+      ? schema
+      : makeExecutableSchema({
+          typeDefs: Array.isArray(typeDefs)
+            ? typeDefs.reduce((prev, next) => prev + '\n' + next)
+            : typeDefs,
+          schemaDirectives,
+          resolvers,
+        });
+
+    this.subscriptions = subscriptions;
+    this.cors = cors;
+
+    if (app) {
+      this.app = app;
```

Comment:
@glasser I was/am intending that but it didn't seem necessary to rework the other methods of this class to not have access to it.

It is not a public API, won't be documented as such, and if people do rely on it, it will be at their own risk